### PR TITLE
Added expiration time to cached data

### PR DESCRIPTION
**Pull request by Tommaso-n**
Default value of expiration time set to 3600 seconds (1 hour).
If set to 0 or negative or None, data will never expire, otherwise set to the provided value in seconds

### CODE REVIEW

**Comment by andreroggeri** [comment]
Hey @Tommaso-n I see that you're still making some changes on this PR. When you're done, please remove the WIP from the PR name so I can review.

**Comment by Tommaso-n** [comment]
> Hey @Tommaso-n I see that you're still making some changes on this PR. When you're done, please remove the WIP from the PR name so I can review.

Hi!
Yes, sorry...now it should be definitive.

I've added a password for redis to the cached configuration (set to None by default), i use Redis Labs and it's required.

I've added two protected properties to the cached client:
- _redis_ttl is the expiration time of cached data, set to 3600 seconds by default.
- _redis_prefix is a string to be added as a prefix to the keys of the cached data for easier retrivial.

I've implemented the setter and getter of _redis_ttl.
If the new value is positive, set the property to the new value. If it's 0, negative or None, set the property to None (cached data never expire).

i've modified the get method to use the prefix for the keys.

To manually remove all the cached data relative to this package from cache, i've implemented a clear_cache method.
It scan for keys beginning with the given prefix and delete them.

This is all :)

File: ynab_sdk/utils/clients/cached_client.py [hunk 1]
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
...
+        else:
+            self.logger.error(f'cache_time_to_live: 0, negative or None TTL value: {ttl_value}, set to None')
+            self._redis_ttl = None
+
+    def clear_cache(self) -> None:
+        keys_count: int = 0
+        for k in self.redis.scan_iter(''.join([self._redis_prefix, '*'])):
+            self.redis.delete(k)
+            keys_count += 1
+        self.logger.error(f'clear_cache: {keys_count} keys deleted from cache')
```
Comment (andreroggeri) [inline ynab_sdk/utils/clients/cached_client.py:37]: Use logger.info since this is not an error

File: ynab_sdk/utils/clients/cached_client.py [hunk 2]
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
...
+        self._redis_prefix: str = 'YNAB_ep_'
+
+    @property
+    def cache_time_to_live(self) -> Union[int, None]:
+        return self._redis_ttl
+
+    @cache_time_to_live.setter
+    def cache_time_to_live(self, ttl_value: Union[int, None]) -> None:
+        if ttl_value and ttl_value > 0:
+            self.logger.error(f'cache_time_to_live: TTL value set to {ttl_value}')
```
Comment (andreroggeri) [inline ynab_sdk/utils/clients/cached_client.py:26]: Use logger.debug since this is not an error

File: ynab_sdk/utils/clients/cached_client.py [hunk 3]
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
...
+            self._redis_ttl = ttl_value
+        else:
+            self.logger.error(f'cache_time_to_live: 0, negative or None TTL value: {ttl_value}, set to None')
```
Comment (andreroggeri) [inline ynab_sdk/utils/clients/cached_client.py:29]: Use logger.debug since this is not an error

File: ynab_sdk/utils/clients/cached_client.py [hunk 4]
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
 
     def __init__(self, config: CachedConfig):
         super().__init__(config)
-        self.redis = Redis(host=config.redis_host, port=config.redis_port, db=config.redis_db)
+        self.redis = Redis(host=config.redis_host, port=config.redis_port, db=config.redis_db, password=config.redis_pass)
+        self._redis_ttl: Union[int, None] = 3600
```
Comment (andreroggeri) [inline ynab_sdk/utils/clients/cached_client.py:16]: Optional[int] its the correct typing here (Has the same meaning, but a matter of semantics)

File: ynab_sdk/utils/configurations/cached.py [hunk 5]
```diff
@@ -3,8 +3,9 @@
 
 class CachedConfig(DefaultConfig):
 
-    def __init__(self, redis_host: str, redis_port: int, redis_db: int = 0, **kwargs):
+    def __init__(self, redis_host: str, redis_port: int, redis_db: int = 0, redis_pass: str = None, **kwargs) -> None:
```
Comment (andreroggeri) [inline ynab_sdk/utils/configurations/cached.py:6]: `redis_pass` should be typed as Optional[str]

File: ynab_sdk/utils/clients/cached_client.py (hunk 4 above)
Comment (andreroggeri) [inline ynab_sdk/utils/clients/cached_client.py:16]: Hmm, maybe the ttl should be a property of the CachedConfig.
Its a little odd that after all configuration I still have to call another config method

File: ynab_sdk/utils/clients/cached_client.py [hunk 6]
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
...
+        self._redis_prefix: str = 'YNAB_ep_'
+
+    @property
+    def cache_time_to_live(self) -> Union[int, None]:
+        return self._redis_ttl
+
+    @cache_time_to_live.setter
+    def cache_time_to_live(self, ttl_value: Union[int, None]) -> None:
```
Comment (andreroggeri) [inline ynab_sdk/utils/clients/cached_client.py:24]: `ttl_value`  should be Optional[int]

File: ynab_sdk/utils/clients/cached_client.py (hunk 1 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/clients/cached_client.py:37]: Done

File: ynab_sdk/utils/clients/cached_client.py (hunk 2 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/clients/cached_client.py:26]: Done

File: ynab_sdk/utils/clients/cached_client.py (hunk 3 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/clients/cached_client.py:29]: Done

File: ynab_sdk/utils/configurations/cached.py (hunk 5 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/configurations/cached.py:6]: Done

File: ynab_sdk/utils/clients/cached_client.py (hunk 4 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/clients/cached_client.py:16]: Done

File: ynab_sdk/utils/clients/cached_client.py (hunk 6 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/clients/cached_client.py:24]: Done

File: ynab_sdk/utils/clients/cached_client.py (hunk 4 above)
Comment (Tommaso-n) [reply ynab_sdk/utils/clients/cached_client.py]: Moved

File: ynab_sdk/utils/configurations/cached.py [hunk 7]
```diff
@@ -1,10 +1,27 @@
...
+    def __init__(self, redis_host: str, redis_port: int, redis_db: int = 0, redis_pass: Optional[str] = None, **kwargs) -> None:
         super().__init__(**kwargs)
         self.redis_host = redis_host
         self.redis_port = redis_port
         self.redis_db = redis_db
+        self.redis_pass = redis_pass
+        self._redis_ttl: Optional[int] = 3600
+
+    @property
+    def redis_ttl(self) -> Union[int, None]:
```
Comment (andreroggeri) [inline ynab_sdk/utils/configurations/cached.py:17]: I missed this one: the return type is Optional[int]

Comment (Tommaso-n) [reply ynab_sdk/utils/configurations/cached.py]: I missed it too...corrected

**Pull request merged**